### ISSUE TITLE
Returning upvar captured by mutable reference doesn't work

### ISSUE THREAD

**Issue by Zoxc** [rust-lang/rust#40510]
This works:
```rust
fn main() {
    let a = Box::new(8u8);

    || {
        &a
    };
}
```

This does not and gives a confusing error:
```rust
fn main() {
    let mut a = Box::new(8u8);

    || {
        &mut a
    };
}
```
The error:
```
error[E0495]: cannot infer an appropriate lifetime for borrow expression due to conflicting requirements
 --> <anon>:5:9
  |
5 |         &mut a
  |         ^^^^^^
  |
note: first, the lifetime cannot outlive the lifetime  as defined on the body at 4:7...
 --> <anon>:4:8
  |
4 |       || {
  |  ________^ starting here...
5 | |         &mut a
6 | |     };
  | |_____^ ...ending here
note: ...so that closure can access `a`
 --> <anon>:5:9
  |
5 |         &mut a
  |         ^^^^^^
note: but, the lifetime must be valid for the scope of call-site for function at 4:7...
 --> <anon>:4:8
  |
4 |       || {
  |  ________^ starting here...
5 | |         &mut a
6 | |     };
  | |_____^ ...ending here
note: ...so that return value is valid for the call
 --> <anon>:4:5
  |
4 |       || {
  |  _____^ starting here...
5 | |         &mut a
6 | |     };
  | |_____^ ...ending here
```

**Comment by Zoxc** [comment]
This also gives a confusing error:
```rust
fn main() {
	let mut a = Vec::<bool>::new();

	|| {
		|| {
			a.push(true)
		}
	};
}
```
The following works:
```rust
fn main() {
	let a = Vec::<bool>::new();

	|| {
		|| {
			a.len()
		}
	};
}
```

**Comment by Mark-Simulacrum** [comment]
None of these examples don't compile today, so I'm presuming this is fixed and marking E-needstest.

**Comment by dsprenkels** [comment]
Can I take this issue and put this into a run-pass test?

**Comment by Mark-Simulacrum** [comment]
Sounds good to me, though compile-pass would probably be sufficient - grep for `rustc_error`.

**Comment by dsprenkels** [comment]
Ok. Will do. Btw, it looks like this commit was fixed by 9809fe0d021ebed29d13a4f7fa58c660107da384.
